### ISSUE TITLE
Add keyboard support for insert table dropdown

### ISSUE THREAD

**Issue by jodator** [ckeditor/ckeditor5#3176]
While implementing insert table feature:

![selection_053](https://user-images.githubusercontent.com/247363/40646701-3f39e1b0-632a-11e8-814a-a2f929f7d1da.png)

I've realized that it has no keyboard support defined. But I'd leave it as a followup to focus first on other missing features from MVP.

I also had some minor issues with getting this to work with `FocusCycler` and using `<div tabindex="-1">`. Also maybe I'm missing something but with current implementation of above dropdown I'd like to use methods like `focus( 10 )` and `focus( -10 )` to skip a whole row on `arrowup` and `arrowdown`.

For clarity this issue is only about properly **handling the keyboard** in table dropdown.

**Comment by oleq** [comment]
Some new ideas (or just confirmation of what we came up with) in this ticket: #6765.

**Comment by dpavez14** [comment]
It seems that the best approach could be something as Comandeer suggested or just replace the grid by 2 inputs, because this way it would be accessible for other kind of input methods, as voice commands (eg. Voice Control, Voice Access).

**Comment by oleq** [comment]
> or just replace the grid by 2 inputs

That would make the UI nearly useless for anyone using mouse so it's definitely a no-go. The grid is one thing and the inputs are another, and I think what @Comandeer suggested is a very good idea in this case.

**Comment by oleq** [comment]
Related #7939.

**Comment by magda-chrzescian** [comment]
_**I have done some research on this and summed up the discussion below:**_

## Context and Problem Statement

The UI for inserting tables should be accessible for as many input methods as possible. 

## Decision Drivers 

* It is impossible to insert a table using a keyboard.
* It is impossible to insert a table using a device with a touch screen.
* It is impossible to insert a table using voice control.

## Considered Options

* Extension of the current interface by adding the keyboard, touch, and voice control support
* Replacing the current interface with two inputs for defining the amounts of rows and columns
* Mix of the two above: Adding the extra inputs and displaying the existing interface only on devices with the pointer support

## Decision Outcome

The 'Mix of the two above' option seems to be the most reasonable.

It would cover the need to support multiple input methods with minimal effort thanks to the additional inputs and will also keep the super fast and simple table insertion for desktop users.

## Pros and Cons of the Options 

### Extension of the current interface by adding the keyboard, touch, and voice control support

The detailed description of this one can be found here: #6765
Plus, the touch support should include enlarging the cell size, adding an 'insert table' button for the touch screen, and adding support for at least mouseDown and mouseUp events in the editor.

👍 a pretty consistent interface on various devices,
👍 the current behavior of the interface doesn't change (comfortable for existing users)
👍 one-click table insertion for mouse users,
👍 very simple interface for keyboard users (no navigation between inputs, just ⬇️ ➡️ and enter,

⚪ sufficient support for touch devices,

❔I'm not sure how this solution would work with the screen readers,

❌ probably a huge effort to implement


### Replacing the current interface with two inputs for defining the amounts of rows and columns

* Tab should be used for navigating between inputs and arrows for increasing and decreasing the values.
* '+' and '-' should not be focusable.
* The editable labels should have extra buttons if we want them to be accessible on the touch devices.
![image.png](https://images.zenhubusercontent.com/5fc63519d774781c638d81af/672797ab-7e89-4df0-99b6-c71e9ca9790f)

👍 a consistent interface on various devices,
👍 relatively simple to implement,
👍 available for screen readers and voice commands,

⚪ sufficient support for keyboard,
⚪ sufficient support for touch devices,

❌ a lot of clicking if a mouse user wants to insert a big table,
❌ user has to learn from the beginning how to use the interface - can be irritating if someone got used to the existing one


### Mix of the two above: Adding the extra inputs and displaying the existing interface only on devices with the pointer support

* Tab should be used for navigation between input, arrows/scroll (native) for increasing and decreasing the values.
* '+' and '-' should not be focusable.
* The existing grid should not be focusable.
* The inputs should have extra buttons if we want them to be accessible on the touch devices.
* Table is added to the editor by clicking the grid, clicking the 'insert' button, and pressing enter while one of the inputs or a button is focused.
* The sizes of the elements and hiding the grid on touch devices can be handled with interaction media features.
![image.png](https://images.zenhubusercontent.com/5fc63519d774781c638d81af/da1885a5-bd13-4890-8cab-1ebba6c740c1)

👍 relatively simple to implement,
👍 available for screen readers and voice commands,
👍 the current behavior of the interface doesn't change (comfortable for existing users)
👍 one-click table insertion for mouse users,

⚪ sufficient support for keyboard,
⚪ sufficient support for touch devices
⚪ a relatively consistent interface on various devices (differences handled with CSS only)

**Comment by Reinmar** [comment]
> ### Replacing the current interface with two inputs for defining the amounts of rows and columns

For me, the grid is very convenient for mouse users and replacing it with number inputs will be a big step back. So, to the list of cons, I'd add the fact that it would negatively affect the current UX.

**Comment by oleq** [comment]
Thanks, @magda-chrzescian for the research. 

I agree with Reinmar that the grid must stay for desktop users and for backward (UX) compatibility.

BTW: Have you considered a UX that would roughly look like this in your research?

![](https://user-images.githubusercontent.com/1099479/103524982-c43b5280-4e7e-11eb-9c88-94aa6ef51cbb.png)

In this approach:

*   the grid works with the mouse as in the current implementation,
*   but you can also focus the inputs to type the dimensions 
    *   when any of the fields is focused a button appears nearby (next to or underneath) allowing to accept the dimensions and insert a new table

**Comment by oleq** [comment]
Also, I'd add "Ability to insert tables bigger than 10x10" to the **Decision Drivers** section. Maybe this is not our top priority, but let's keep it in mind.

**Comment by magda-chrzescian** [comment]
@oleq yes, I considered this option. I rejected it because of the need to support touch devices - it doesn't really look like something that can be edited and the inputs would be too small to focus properly. I proposed the +/- buttons as the extension of this idea, which would be more convenient for touch screen users.  
  
But, after rethinking, I believe that it can be used in UI for touch devices under some conditions:

*   the size of the inputs must be enlarged enough to be focusable with the finger (usually ~40px),
*   the first input must be auto-focused after a user opens the dropdown, so he can discover that it's editable

And I still see a relevant inconvenience - I'm not sure if it is obvious which input reflects the number of columns and which the number of rows. It's going to be clear only after entering the value and focusing the second input (the grid also must reflect the state of the inputs).  
  
I will summarize our thoughts and add another option to the ADR.

**Comment by magda-chrzescian** [comment]
I updated the ADR, according to our earlier conversation with @oleq.

\* I updated the solution for additional inputs, so the dropdown is smaller in the default state.

\* I added the proposed option with editable labels

\* Removed the option, where the grid was removed completely (kept it on the list in the case we would like to go back to it)

**Comment by mlewand** [comment]
The column/row count input has been extracted to a dedicated issue #11912 as it's a separate thing.

**Comment by mlewand** [comment]
This issue should be solved after we figure out how we want to handle grids in #3176.

Basically the idea here is to treat this grid as any other grid with following exception:

*   As focused / hovered grid item changes the highlight should follow.
*   As focused / hovered grid item changes the column/row number on the bottom should be updated accordingly.

We should also check RTL languages (because in this case the highlighting should go to bottom **left** corner).

**Comment by wimleers** [comment]
FYI: tracking this at https://www.drupal.org/project/drupal/issues/3283801 😊

**Comment by mlewand** [comment]
#3184 marked as a (partial) duplicate of this task.
